### PR TITLE
Bring your own logger

### PR DESCRIPTION
**Pull request by sylr**
This PR removes logrus (except for the tests).

It gives the ability to anybody using oxy to provide it with a custom logger as long as it satisfies the following interface:

```go
type Logger interface {
	Debugf(string, ...interface{})
	Infof(string, ...interface{})
	Warnf(string, ...interface{})
	Errorf(string, ...interface{})
	Fatalf(string, ...interface{})
}
```

This interface satisfies `github.com/sirupsen/logrus.Logger` and `go.uber.org/zap.SugaredLogger`.

As there is a few signatures change the module path as been changed from `github.com/vulcand/oxy` to `github.com/vulcand/oxy/v2`.

There is also a behaviour change as there is no logging done if no logger is provided.

Fixes https://github.com/vulcand/oxy/issues/202 https://github.com/vulcand/oxy/issues/83

### CODE REVIEW

File: buffer/buffer.go [hunk 1]
```diff
@@ -209,14 +217,14 @@ func (b *Buffer) Wrap(next http.Handler) error {
 }
 
 func (b *Buffer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	if b.log.Level >= log.DebugLevel {
-		logEntry := b.log.WithField("Request", utils.DumpHttpRequest(req))
-		logEntry.Debug("vulcand/oxy/buffer: begin ServeHttp on request")
-		defer logEntry.Debug("vulcand/oxy/buffer: completed ServeHttp on request")
+	if b.debug() {
+		dumb := utils.DumpHttpRequest(req)
```
Comment (dustin-decker) [inline buffer/buffer.go:221]: dumb->dump

**Comment by dustin-decker** [comment]
I think logger customization is definitely needed, but couldn't it be done without breaking changes by having the Logger setter and defaulting to the current logger? Also, why have the debug logger func instead of relying on leveled logging?

**Comment by sylr** [comment]
The debug func is here to prevent calling `utils.DumpHttpRequest(req)` when we know the output is going to be dropped.

When using only logrus we could just test for the current loglevel but now we are logger agnostic we need something else, hence the LoggerDebugFunc. If you have a better idea I'm game.

File: buffer/buffer.go (hunk 1 above)
Comment (sylr) [reply buffer/buffer.go]: Thanks

**Comment by sylr** [comment]
Of course we could also just drop the logging of the request dump if this is acceptable.

**Pull request closed without merging**